core/consensus/aura/src/lib.rs

```diff
@@ -154,247 +157,201 @@ pub fn start_aura_thread<B, C, E, I, SO, Error>(
-							"Unable to fetch authorities at block {:?}: {:?}",
-							chain_head.hash(),
-							e
-						);
-						return Either::B(future::ok(()));
-					}
-				};
-
-				let proposal_work = match slot_author(slot_num, &authorities) {
-					None => return Either::B(future::ok(())),
-					Some(author) => if author.0 == public_key.0 {
-						debug!(
-							target: "aura", "Starting authorship at slot {}; timestamp = {}",
-							slot_num,
-							timestamp
-						);
+	let worker = AuraWorker {
+		client: client.clone(), block_import, env, local_key
+	};
 
-						// we are the slot author. make a block and sign it.
-						let proposer = match env.init(&chain_head, &authorities) {
-							Ok(p) => p,
-							Err(e) => {
-								warn!("Unable to author block in slot {:?}: {:?}", slot_num, e);
-								return Either::B(future::ok(()))
-							}
-						};
-
-						let remaining_duration = slot_info.remaining_duration();
-						// deadline our production to approx. the end of the
-						// slot
-						Timeout::new(
-							proposer.propose(slot_info.inherent_data).into_future(),
-							remaining_duration,
-						)
-					} else {
-						return Either::B(future::ok(()));
-					}
-				};
+	aura_slots::start_slot_worker::<_, _, _, _, AuraSlotCompatible, _>(
+		slot_duration,
+		client,
+		Arc::new(worker),
+		sync_oracle,
+		on_exit,
+		inherent_data_providers
+	)
+}
 
-				let block_import = block_import.clone();
-				Either::A(proposal_work
-					.map(move |b| {
-						// minor hack since we don't have access to the timestamp
-						// that is actually set by the proposer.
-						let slot_after_building = slot_now(slot_duration);
-						if slot_after_building != Some(slot_num) {
-							info!(
-								"Discarding proposal for slot {}; block production took too long",
-								slot_num
-							);
-							return
-						}
+struct AuraWorker<C, E, I> {
+	client: Arc<C>,
+	block_import: Arc<I>,
+	env: Arc<E>,
+	local_key: Arc<ed25519::Pair>,
+}
 
-						let (header, body) = b.deconstruct();
-						let header_num = header.number().clone();
-						let pre_hash = header.hash();
-						let parent_hash = header.parent_hash().clone();
-
-						// sign the pre-sealed hash of the block and then
-						// add it to a digest item.
-						let to_sign = (slot_num, pre_hash).encode();
-						let signature = pair.sign(&to_sign[..]);
-						let item = <DigestItemFor<B> as CompatibleDigestItem>::aura_seal(
-							slot_num,
-							signature,
-						);
+impl<B: Block, C, E, I, Error> SlotWorker<B> for AuraWorker<C, E, I> where
+	C: Authorities<B>,
+	E: Environment<B, Error=Error>,
+	E::Proposer: Proposer<B, Error=Error> + 'static,
+	I: BlockImport<B> + 'static,
+	Error: From<C::Error> + From<I::Error>,
+	DigestItemFor<B>: CompatibleDigestItem + DigestItem<AuthorityId=Ed25519AuthorityId>,
+	Error: ::std::error::Error + Send + 'static + From<::consensus_common::Error>,
+{
+	fn on_start(
+		&self,
+		inherent_data_providers: &InherentDataProviders,
+		slot_duration: u64
```

Comment:
Shouldn't we use the `SlotDuration` struct here?

Reply:
I think the reason for `SlotDuration` type is that it manages fetching the actual value from client? But here we don't need that, only the raw value.

Reply:
I thought that we have this type to uniquely identify the `SlotDuration` as type.

Reply:
In other places like `Slots` we also only use the raw `u64` type.